File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/InAppMessageManagerBase.java
package com.braze.ui.inappmessage;

import android.app.Activity;
import android.content.Context;

import androidx.annotation.Nullable;
import androidx.annotation.RestrictTo;

import com.braze.models.inappmessage.IInAppMessage;
import com.braze.support.BrazeLogger;
import com.braze.ui.inappmessage.factories.DefaultInAppMessageAnimationFactory;
import com.braze.ui.inappmessage.factories.DefaultInAppMessageFullViewFactory;
import com.braze.ui.inappmessage.factories.DefaultInAppMessageHtmlFullViewFactory;
import com.braze.ui.inappmessage.factories.DefaultInAppMessageHtmlViewFactory;
import com.braze.ui.inappmessage.factories.DefaultInAppMessageModalViewFactory;
import com.braze.ui.inappmessage.factories.DefaultInAppMessageSlideupViewFactory;
import com.braze.ui.inappmessage.factories.DefaultInAppMessageViewWrapperFactory;
import com.braze.ui.inappmessage.listeners.DefaultHtmlInAppMessageActionListener;
import com.braze.ui.inappmessage.listeners.DefaultInAppMessageManagerListener;
import com.braze.ui.inappmessage.listeners.DefaultInAppMessageWebViewClientListener;
import com.braze.ui.inappmessage.listeners.IHtmlInAppMessageActionListener;
import com.braze.ui.inappmessage.listeners.IInAppMessageManagerListener;
import com.braze.ui.inappmessage.listeners.IInAppMessageWebViewClientListener;

public class InAppMessageManagerBase {
  private static final String TAG = BrazeLogger.getBrazeLogTag(InAppMessageManagerBase.class);

  private boolean mClickOutsideModalDismissesInAppMessageView = false;
  private boolean mBackButtonDismissesInAppMessageView = true;
  @Nullable
  Activity mActivity;
  @Nullable
  Context mApplicationContext;

  // view listeners
  private final IInAppMessageWebViewClientListener mInAppMessageWebViewClientListener = new DefaultInAppMessageWebViewClientListener();

  // html action listeners
  private final IHtmlInAppMessageActionListener mDefaultHtmlInAppMessageActionListener = new DefaultHtmlInAppMessageActionListener();

  // factories
  private final IInAppMessageViewFactory mInAppMessageSlideupViewFactory = new DefaultInAppMessageSlideupViewFactory();
  private final IInAppMessageViewFactory mInAppMessageModalViewFactory = new DefaultInAppMessageModalViewFactory();
  private final IInAppMessageViewFactory mInAppMessageFullViewFactory = new DefaultInAppMessageFullViewFactory();
  private final IInAppMessageViewFactory mInAppMessageHtmlFullViewFactory = new DefaultInAppMessageHtmlFullViewFactory(mInAppMessageWebViewClientListener);
  private final IInAppMessageViewFactory mInAppMessageHtmlViewFactory = new DefaultInAppMessageHtmlViewFactory(mInAppMessageWebViewClientListener);

  // animation factory
  private final IInAppMessageAnimationFactory mInAppMessageAnimationFactory = new DefaultInAppMessageAnimationFactory();

  // manager listeners
  private final IInAppMessageManagerListener mDefaultInAppMessageManagerListener = new DefaultInAppMessageManagerListener();

  // view wrapper factory
  private final IInAppMessageViewWrapperFactory mDefaultInAppMessageViewWrapperFactory = new DefaultInAppMessageViewWrapperFactory();

  // custom listeners
  @Nullable
  private IInAppMessageViewFactory mCustomInAppMessageViewFactory;
  @Nullable
  private IInAppMessageAnimationFactory mCustomInAppMessageAnimationFactory;
  @Nullable
  private IInAppMessageManagerListener mCustomInAppMessageManagerListener;
  @Nullable
  private IInAppMessageViewWrapperFactory mCustomInAppMessageViewWrapperFactory;
  @Nullable
  private IHtmlInAppMessageActionListener mCustomHtmlInAppMessageActionListener;
  /**
   * A custom listener to be fired for control in-app messages.
   * <p/>
   * see {@link IInAppMessage#isControl()}
   */
  @Nullable
  private IInAppMessageManagerListener mCustomControlInAppMessageManagerListener;

  public IInAppMessageManagerListener getInAppMessageManagerListener() {
    return mCustomInAppMessageManagerListener != null ? mCustomInAppMessageManagerListener : mDefaultInAppMessageManagerListener;
  }

  /**
   * A {@link IInAppMessageManagerListener} to be used only for control in-app messages.
   * <p/>
   * see {@link IInAppMessage#isControl()}
   */
  public IInAppMessageManagerListener getControlInAppMessageManagerListener() {
    return mCustomControlInAppMessageManagerListener != null ? mCustomControlInAppMessageManagerListener : mDefaultInAppMessageManagerListener;
  }

  public IHtmlInAppMessageActionListener getHtmlInAppMessageActionListener() {
    return mCustomHtmlInAppMessageActionListener != null ? mCustomHtmlInAppMessageActionListener : mDefaultHtmlInAppMessageActionListener;
  }

  /**
   * Gets the default {@link IInAppMessageViewFactory} as returned by the {@link BrazeInAppMessageManager}
   * for the given {@link IInAppMessage}.
   *
   * @return The {@link IInAppMessageViewFactory} or null if the message type does not have a {@link IInAppMessageViewFactory}.
   */
  public IInAppMessageViewFactory getDefaultInAppMessageViewFactory(IInAppMessage inAppMessage) {
    switch (inAppMessage.getMessageType()) {
      case SLIDEUP:
        return mInAppMessageSlideupViewFactory;
      case MODAL:
        return mInAppMessageModalViewFactory;
      case FULL:
        return mInAppMessageFullViewFactory;
      case HTML_FULL:
        return mInAppMessageHtmlFullViewFactory;
      case HTML:
        return mInAppMessageHtmlViewFactory;
      default:
        BrazeLogger.w(TAG, "Failed to find view factory for in-app message with type: " + inAppMessage.getMessageType());
        return null;
    }
  }

  public IInAppMessageViewWrapperFactory getInAppMessageViewWrapperFactory() {
    return mCustomInAppMessageViewWrapperFactory != null ? mCustomInAppMessageViewWrapperFactory : mDefaultInAppMessageViewWrapperFactory;
  }

  public boolean getDoesBackButtonDismissInAppMessageView() {
    return mBackButtonDismissesInAppMessageView;
  }

  public boolean getDoesClickOutsideModalViewDismissInAppMessageView() {
    return mClickOutsideModalDismissesInAppMessageView;
  }

  public IInAppMessageAnimationFactory getInAppMessageAnimationFactory() {
    return mCustomInAppMessageAnimationFactory != null ? mCustomInAppMessageAnimationFactory : mInAppMessageAnimationFactory;
  }

  public IInAppMessageViewFactory getInAppMessageViewFactory(IInAppMessage inAppMessage) {
    if (mCustomInAppMessageViewFactory != null) {
      return mCustomInAppMessageViewFactory;
    } else {
      return getDefaultInAppMessageViewFactory(inAppMessage);
    }
  }

  @Nullable
  public Activity getActivity() {
    return mActivity;
  }

  @Nullable
  public Context getApplicationContext() {
    return mApplicationContext;
  }

  /**
   * Sets whether the hardware back button dismisses in-app messages. Defaults to true.
   * Note that the hardware back button default behavior will be used instead (i.e. the host {@link Activity}'s
   * {@link Activity#onKeyDown(int, KeyEvent)} method will be called).
   */
  public void setBackButtonDismissesInAppMessageView(boolean backButtonDismissesInAppMessageView) {
    BrazeLogger.d(TAG, "In-App Message back button dismissal set to " + backButtonDismissesInAppMessageView);
    mBackButtonDismissesInAppMessageView = backButtonDismissesInAppMessageView;
  }

  /**
   * Sets whether the tapping outside the modal in-app message content dismiss the
   * message. Defaults to false.
   */
  public void setClickOutsideModalViewDismissInAppMessageView(boolean doesDismiss) {
    BrazeLogger.d(TAG, "Modal In-App Message outside tap dismissal set to " + doesDismiss);
    mClickOutsideModalDismissesInAppMessageView = doesDismiss;
  }

  /**
   * Assigns a custom {@link IInAppMessageManagerListener} that will be used when displaying in-app messages. To revert
   * back to the default {@link IInAppMessageManagerListener}, call this method with null.
   * <p/>
   * see {@link IInAppMessage#isControl()}
   *
   * @param inAppMessageManagerListener A custom {@link IInAppMessageManagerListener} or null (to revert back to the
   *                                    default {@link IInAppMessageManagerListener}).
   */
  public void setCustomInAppMessageManagerListener(IInAppMessageManagerListener inAppMessageManagerListener) {
    BrazeLogger.d(TAG, "Custom InAppMessageManagerListener set");
    mCustomInAppMessageManagerListener = inAppMessageManagerListener;
  }

  /**
   * Assigns a custom {@link IInAppMessageManagerListener} that will be used when displaying control in-app messages. To revert
   * back to the default {@link IInAppMessageManagerListener}, call this method with null.
   *
   * @param inAppMessageManagerListener A custom {@link IInAppMessageManagerListener} for control in-app messages or null (to revert back to the
   *                                    default {@link IInAppMessageManagerListener}).
   */
  public void setCustomControlInAppMessageManagerListener(IInAppMessageManagerListener inAppMessageManagerListener) {
    BrazeLogger.d(TAG, "Custom ControlInAppMessageManagerListener set. This listener will only be used for control in-app messages.");
    mCustomControlInAppMessageManagerListener = inAppMessageManagerListener;
  }

  /**
   * Assigns a custom IHtmlInAppMessageActionListener that will be used during the display of Html in-app messages.
   *
   * @param htmlInAppMessageActionListener A custom IHtmlInAppMessageActionListener or null (to revert back to the
   *                                       default IHtmlInAppMessageActionListener).
   */
  public void setCustomHtmlInAppMessageActionListener(IHtmlInAppMessageActionListener htmlInAppMessageActionListener) {
    BrazeLogger.d(TAG, "Custom htmlInAppMessageActionListener set");
    mCustomHtmlInAppMessageActionListener = htmlInAppMessageActionListener;
  }

  /**
   * Assigns a custom IInAppMessageAnimationFactory that will be used to animate the in-app message View. To revert
   * back to the default IInAppMessageAnimationFactory, call the setCustomInAppMessageAnimationFactory method with null.
   *
   * @param inAppMessageAnimationFactory A custom IInAppMessageAnimationFactory or null (to revert back to the default
   *                                     IInAppMessageAnimationFactory).
   */
  public void setCustomInAppMessageAnimationFactory(IInAppMessageAnimationFactory inAppMessageAnimationFactory) {
    BrazeLogger.d(TAG, "Custom InAppMessageAnimationFactory set");
    mCustomInAppMessageAnimationFactory = inAppMessageAnimationFactory;
  }

  /**
   * Assigns a custom IInAppMessageViewFactory that will be used to create the in-app message View. To revert
   * back to the default IInAppMessageViewFactory, call the setCustomInAppMessageViewFactory method with null.
   *
   * @param inAppMessageViewFactory A custom IInAppMessageViewFactory or null (to revert back to the default
   *                                IInAppMessageViewFactory).
   */
  public void setCustomInAppMessageViewFactory(IInAppMessageViewFactory inAppMessageViewFactory) {
    BrazeLogger.d(TAG, "Custom InAppMessageViewFactory set");
    mCustomInAppMessageViewFactory = inAppMessageViewFactory;
  }

  /**
   * Sets a custom {@link IInAppMessageViewWrapperFactory} that will be used to
   * display an {@link IInAppMessage} to the user.
   */
  public void setCustomInAppMessageViewWrapperFactory(@Nullable IInAppMessageViewWrapperFactory inAppMessageViewWrapperFactory) {
    BrazeLogger.d(TAG, "Custom IInAppMessageViewWrapperFactory set");
    mCustomInAppMessageViewWrapperFactory = inAppMessageViewWrapperFactory;
  }

  @RestrictTo(RestrictTo.Scope.TESTS)
  public Boolean isActivitySet() {
    return mActivity != null;
  }
}

File: android-sdk-ui/src/main/java/com/braze/ui/inappmessage/InAppMessageManagerBase.kt
package com.braze.ui.inappmessage

import android.app.Activity
import android.content.Context
import androidx.annotation.RestrictTo
import com.braze.enums.inappmessage.MessageType
import com.braze.models.inappmessage.IInAppMessage
import com.braze.support.BrazeLogger.Priority.W
import com.braze.support.BrazeLogger.brazelog
import com.braze.ui.inappmessage.factories.DefaultInAppMessageAnimationFactory
import com.braze.ui.inappmessage.factories.DefaultInAppMessageFullViewFactory
import com.braze.ui.inappmessage.factories.DefaultInAppMessageHtmlFullViewFactory
import com.braze.ui.inappmessage.factories.DefaultInAppMessageHtmlViewFactory
import com.braze.ui.inappmessage.factories.DefaultInAppMessageModalViewFactory
import com.braze.ui.inappmessage.factories.DefaultInAppMessageSlideupViewFactory
import com.braze.ui.inappmessage.factories.DefaultInAppMessageViewWrapperFactory
import com.braze.ui.inappmessage.listeners.DefaultHtmlInAppMessageActionListener
import com.braze.ui.inappmessage.listeners.DefaultInAppMessageManagerListener
import com.braze.ui.inappmessage.listeners.DefaultInAppMessageWebViewClientListener
import com.braze.ui.inappmessage.listeners.IHtmlInAppMessageActionListener
import com.braze.ui.inappmessage.listeners.IInAppMessageManagerListener
import com.braze.ui.inappmessage.listeners.IInAppMessageWebViewClientListener

@Suppress("TooManyFunctions")
open class InAppMessageManagerBase {
    open val doesClickOutsideModalViewDismissInAppMessageView
        get() = doesClickOutsideModalViewDismissInAppMessageViewField

    private var doesClickOutsideModalViewDismissInAppMessageViewField = false

    /**
     * Determines whether the next call to
     * [BrazeInAppMessageManager.unregisterInAppMessageManager] will be ignored.
     */
    open var shouldNextUnregisterBeSkipped = false
        set(shouldSkip) {
            brazelog { "Setting setShouldNextUnregisterBeSkipped to $shouldSkip" }
            field = shouldSkip
        }

    open val doesBackButtonDismissInAppMessageView
        get() = doesBackButtonDismissInAppMessageViewField

    private var doesBackButtonDismissInAppMessageViewField = true

    // Since many clients have written code against this, we want to maintain backwards compatibility
    // as much as possible. That's why we have these names and present them as JvmFields.
    @JvmField
    @Suppress("VariableNaming")
    protected var mActivity: Activity? = null

    @JvmField
    @Suppress("VariableNaming")
    protected var mApplicationContext: Context? = null

    // These serve the purpose of allowing people to write more Kotlin-friendly code, but also provide
    // the getActivity() and getApplicationContext() Java generated code for backwards compatibility
    open val activity
        get() = mActivity

    open val applicationContext
        get() = mApplicationContext

    // view listeners
    private val inAppMessageWebViewClientListener: IInAppMessageWebViewClientListener =
        DefaultInAppMessageWebViewClientListener()

    // html action listeners
    private val defaultHtmlInAppMessageActionListener: IHtmlInAppMessageActionListener =
        DefaultHtmlInAppMessageActionListener()

    // factories
    private val inAppMessageSlideupViewFactory: IInAppMessageViewFactory =
        DefaultInAppMessageSlideupViewFactory()
    private val inAppMessageModalViewFactory: IInAppMessageViewFactory =
        DefaultInAppMessageModalViewFactory()
    private val inAppMessageFullViewFactory: IInAppMessageViewFactory =
        DefaultInAppMessageFullViewFactory()
    private val inAppMessageHtmlFullViewFactory: IInAppMessageViewFactory =
        DefaultInAppMessageHtmlFullViewFactory(inAppMessageWebViewClientListener)
    private val inAppMessageHtmlViewFactory: IInAppMessageViewFactory =
        DefaultInAppMessageHtmlViewFactory(inAppMessageWebViewClientListener)

    // animation factory
    private val inAppMessageAnimationFactoryField: IInAppMessageAnimationFactory =
        DefaultInAppMessageAnimationFactory()

    // manager listeners
    private val defaultInAppMessageManagerListener: IInAppMessageManagerListener =
        DefaultInAppMessageManagerListener()

    // view wrapper factory
    private val defaultInAppMessageViewWrapperFactory: IInAppMessageViewWrapperFactory =
        DefaultInAppMessageViewWrapperFactory()

    // custom listeners
    private var customInAppMessageViewFactory: IInAppMessageViewFactory? = null
    private var customInAppMessageAnimationFactory: IInAppMessageAnimationFactory? = null
    private var customInAppMessageManagerListener: IInAppMessageManagerListener? = null
    private var customInAppMessageViewWrapperFactory: IInAppMessageViewWrapperFactory? = null
    private var customHtmlInAppMessageActionListener: IHtmlInAppMessageActionListener? = null

    /**
     * A custom listener to be fired for control in-app messages.
     *
     *
     * see [IInAppMessage.isControl]
     */
    private var customControlInAppMessageManagerListener: IInAppMessageManagerListener? = null
    open val inAppMessageManagerListener: IInAppMessageManagerListener
        get() = customInAppMessageManagerListener ?: defaultInAppMessageManagerListener

    /**
     * A [IInAppMessageManagerListener] to be used only for control in-app messages.
     *
     * see [IInAppMessage.isControl]
     */
    open val controlInAppMessageManagerListener: IInAppMessageManagerListener
        get() = customControlInAppMessageManagerListener ?: defaultInAppMessageManagerListener
    open val htmlInAppMessageActionListener: IHtmlInAppMessageActionListener
        get() = customHtmlInAppMessageActionListener ?: defaultHtmlInAppMessageActionListener

    open val inAppMessageViewWrapperFactory: IInAppMessageViewWrapperFactory
        get() = customInAppMessageViewWrapperFactory ?: defaultInAppMessageViewWrapperFactory
    open val inAppMessageAnimationFactory: IInAppMessageAnimationFactory
        get() = customInAppMessageAnimationFactory ?: inAppMessageAnimationFactoryField

    @get:RestrictTo(RestrictTo.Scope.TESTS)
    open val isActivitySet: Boolean
        get() = activity != null

    /**
     * Gets the default [IInAppMessageViewFactory] as returned by the [BrazeInAppMessageManager]
     * for the given [IInAppMessage].
     *
     * @return The [IInAppMessageViewFactory] or null if the message type does not have a [IInAppMessageViewFactory].
     */
    open fun getDefaultInAppMessageViewFactory(inAppMessage: IInAppMessage): IInAppMessageViewFactory? {
        return when (inAppMessage.messageType) {
            MessageType.SLIDEUP -> inAppMessageSlideupViewFactory
            MessageType.MODAL -> inAppMessageModalViewFactory
            MessageType.FULL -> inAppMessageFullViewFactory
            MessageType.HTML_FULL -> inAppMessageHtmlFullViewFactory
            MessageType.HTML -> inAppMessageHtmlViewFactory
            else -> {
                brazelog(W) {
                    "Failed to find view factory for in-app message with type: ${inAppMessage.messageType}"
                }
                null
            }
        }
    }

    open fun getInAppMessageViewFactory(inAppMessage: IInAppMessage): IInAppMessageViewFactory? =
        customInAppMessageViewFactory ?: getDefaultInAppMessageViewFactory(inAppMessage)

    /**
     * Sets whether the hardware back button dismisses in-app messages. Defaults to true.
     * Note that the hardware back button default behavior will be used instead (i.e. the host [Activity]'s
     * [Activity.onKeyDown] method will be called).
     */
    open fun setBackButtonDismissesInAppMessageView(backButtonDismissesInAppMessageView: Boolean) {
        brazelog { "In-App Message back button dismissal set to $backButtonDismissesInAppMessageView" }
        doesBackButtonDismissInAppMessageViewField = backButtonDismissesInAppMessageView
    }

    /**
     * Sets whether the tapping outside the modal in-app message content dismiss the
     * message. Defaults to false.
     */
    open fun setClickOutsideModalViewDismissInAppMessageView(doesDismiss: Boolean) {
        brazelog { "Modal In-App Message outside tap dismissal set to $doesDismiss" }
        doesClickOutsideModalViewDismissInAppMessageViewField = doesDismiss
    }

    /**
     * Assigns a custom [IInAppMessageManagerListener] that will be used when displaying in-app messages. To revert
     * back to the default [IInAppMessageManagerListener], call this method with null.
     *
     *
     * see [IInAppMessage.isControl]
     *
     * @param inAppMessageManagerListener A custom [IInAppMessageManagerListener] or null (to revert back to the
     * default [IInAppMessageManagerListener]).
     */
    open fun setCustomInAppMessageManagerListener(inAppMessageManagerListener: IInAppMessageManagerListener?) {
        brazelog { "Custom InAppMessageManagerListener set" }
        customInAppMessageManagerListener = inAppMessageManagerListener
    }

    /**
     * Assigns a custom [IInAppMessageManagerListener] that will be used when displaying control in-app messages. To revert
     * back to the default [IInAppMessageManagerListener], call this method with null.
     *
     * @param inAppMessageManagerListener A custom [IInAppMessageManagerListener] for control in-app messages or null (to revert back to the
     * default [IInAppMessageManagerListener]).
     */
    open fun setCustomControlInAppMessageManagerListener(inAppMessageManagerListener: IInAppMessageManagerListener?) {
        brazelog {
            "Custom ControlInAppMessageManagerListener set. This listener will only be used for control in-app messages."
        }
        customControlInAppMessageManagerListener = inAppMessageManagerListener
    }

    /**
     * Assigns a custom IHtmlInAppMessageActionListener that will be used during the display of Html in-app messages.
     *
     * @param htmlInAppMessageActionListener A custom IHtmlInAppMessageActionListener or null (to revert back to the
     * default IHtmlInAppMessageActionListener).
     */
    open fun setCustomHtmlInAppMessageActionListener(htmlInAppMessageActionListener: IHtmlInAppMessageActionListener?) {
        brazelog { "Custom htmlInAppMessageActionListener set" }
        customHtmlInAppMessageActionListener = htmlInAppMessageActionListener
    }

    /**
     * Assigns a custom IInAppMessageAnimationFactory that will be used to animate the in-app message View. To revert
     * back to the default IInAppMessageAnimationFactory, call the setCustomInAppMessageAnimationFactory method with null.
     *
     * @param inAppMessageAnimationFactory A custom IInAppMessageAnimationFactory or null (to revert back to the default
     * IInAppMessageAnimationFactory).
     */
    open fun setCustomInAppMessageAnimationFactory(inAppMessageAnimationFactory: IInAppMessageAnimationFactory?) {
        brazelog { "Custom InAppMessageAnimationFactory set" }
        customInAppMessageAnimationFactory = inAppMessageAnimationFactory
    }

    /**
     * Assigns a custom IInAppMessageViewFactory that will be used to create the in-app message View. To revert
     * back to the default IInAppMessageViewFactory, call the setCustomInAppMessageViewFactory method with null.
     *
     * @param inAppMessageViewFactory A custom IInAppMessageViewFactory or null (to revert back to the default
     * IInAppMessageViewFactory).
     */
    open fun setCustomInAppMessageViewFactory(inAppMessageViewFactory: IInAppMessageViewFactory?) {
        brazelog { "Custom InAppMessageViewFactory set" }
        customInAppMessageViewFactory = inAppMessageViewFactory
    }

    /**
     * Sets a custom [IInAppMessageViewWrapperFactory] that will be used to
     * display an [IInAppMessage] to the user.
     */
    open fun setCustomInAppMessageViewWrapperFactory(inAppMessageViewWrapperFactory: IInAppMessageViewWrapperFactory?) {
        brazelog { "Custom IInAppMessageViewWrapperFactory set" }
        customInAppMessageViewWrapperFactory = inAppMessageViewWrapperFactory
    }
}
